datafusion/expr/src/utils.rs

```diff
@@ -478,20 +479,33 @@ pub fn from_plan(
         }) => {
             let schema =
                 build_join_schema(inputs[0].schema(), inputs[1].schema(), join_type)?;
+
+            let equi_expr_count = on.len();
+            // The preceding part of expr is equi-exprs,
+            // and the struct of each equi-expr is like `left-expr = right-expr`.
+            let new_on:Vec<(Expr,Expr)> = expr.iter().take(equi_expr_count).map(|equi_expr| {
```

Comment:
I think we should error here if `expr` does not has at least `equi_expr_count` elements left. Otherwise I think `take` will silently return fewer than `equi_expr_count` elements, which might result in quite hard to track down bugs

https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.take

Reply:
Make sense, I added a check -- `assert!(expr.len() >= equi_expr_count)`.